// @nuxt/server/dist/server.js
/*!
 * @nuxt/server v2.8.1 (c) 2016-2019

 * - All the amazing contributors
 * Released under the MIT License.
 * Website: https://nuxtjs.org
*/
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

const path = _interopDefault(require('path'));
const consola = _interopDefault(require('consola'));
const launchMiddleware = _interopDefault(require('launch-editor-middleware'));
const serveStatic = _interopDefault(require('serve-static'));
const servePlaceholder = _interopDefault(require('serve-placeholder'));
const connect = _interopDefault(require('connect'));
const utils = require('@nuxt/utils');
const generateETag = _interopDefault(require('etag'));
const fresh = _interopDefault(require('fresh'));
const fs = _interopDefault(require('fs-extra'));
const Youch = _interopDefault(require('@nuxtjs/youch'));
const http = _interopDefault(require('http'));
const https = _interopDefault(require('https'));
const enableDestroy = _interopDefault(require('server-destroy'));
const ip = _interopDefault(require('ip'));
const pify = _interopDefault(require('pify'));
const onHeaders = _interopDefault(require('on-headers'));

class ServerContext {
  constructor(server) {
    this.nuxt = server.nuxt;
    this.globals = server.globals;
    this.options = server.options;
    this.resources = server.resources;
  }
}

async function renderAndGetWindow(
  url = 'http://localhost:3000',
  jsdomOpts = {},
  {
    loadedCallback,
    loadingTimeout = 2000,
    ssr,
    globals
  } = {}
) {
  const jsdom = await Promise.resolve(require('jsdom'))
    .then(m => m.default || m)
    .catch((e) => {
      consola.error(`
         jsdom is not installed. Please install jsdom with:
          $ yarn add --dev jsdom
          OR
          $ npm install --dev jsdom
        `);
      throw e
    });

  const options = Object.assign({
    // Load subresources (https://github.com/tmpvar/jsdom#loading-subresources)
    resources: 'usable',
    runScripts: 'dangerously',
    virtualConsole: true,
    beforeParse(window) {
      // Mock window.scrollTo
      window.scrollTo = () => {};
    }
  }, jsdomOpts);

  const jsdomErrHandler = (err) => {
    throw err
  };

  if (options.virtualConsole) {
    if (options.virtualConsole === true) {
      options.virtualConsole = new jsdom.VirtualConsole().sendTo(consola);
    }
    // Throw error when window creation failed
    options.virtualConsole.on('jsdomError', jsdomErrHandler);
  }

  const { window } = await jsdom.JSDOM.fromURL(url, options);

  // If Nuxt could not be loaded (error from the server-side)
  const nuxtExists = window.document.body.innerHTML.includes(
    ssr ? `window.${globals.context}` : `<div id="${globals.id}">`
  );

  if (!nuxtExists) {
    const error = new Error('Could not load the nuxt app');
    error.body = window.document.body.innerHTML;
    window.close();
    throw error
  }

  // Used by Nuxt.js to say when the components are loaded and the app ready
  await utils.timeout(new Promise((resolve) => {
    window[loadedCallback] = () => resolve(window);
  }), loadingTimeout, `Components loading in renderAndGetWindow was not completed in ${loadingTimeout / 1000}s`);

  if (options.virtualConsole) {
    // After window initialized successfully
    options.virtualConsole.removeListener('jsdomError', jsdomErrHandler);
  }

  // Send back window object
  return window
}

const nuxtMiddleware = ({ options, nuxt, renderRoute, resources }) => async function nuxtMiddleware(req, res, next) {
  // Get context
  const context = utils.getContext(req, res);

  try {
    const url = decodeURI(req.url);
    res.statusCode = 200;
    const result = await renderRoute(url, context);

    // If result is falsy, call renderLoading
    if (!result) {
      await nuxt.callHook('server:nuxt:renderLoading', req, res);
      return
    }

    await nuxt.callHook('render:route', url, result, context);
    const {
      html,
      cspScriptSrcHashes,
      error,
      redirected,
      preloadFiles
    } = result;

    if (redirected) {
      await nuxt.callHook('render:routeDone', url, result, context);
      return html
    }
    if (error) {
      res.statusCode = context.nuxt.error.statusCode || 500;
    }

    // Add ETag header
    if (!error && options.render.etag) {
      const etag = generateETag(html, options.render.etag);
      if (fresh(req.headers, { etag })) {
        res.statusCode = 304;
        res.end();
        await nuxt.callHook('render:routeDone', url, result, context);
        return
      }
      res.setHeader('ETag', etag);
    }

    // HTTP2 push headers for preload assets
    if (!error && options.render.http2.push) {
      // Parse resourceHints to extract HTTP.2 prefetch/push headers
      // https://w3c.github.io/preload/#server-push-http-2
      const { shouldPush, pushAssets } = options.render.http2;
      const { publicPath } = resources.clientManifest;

      const links = pushAssets
        ? pushAssets(req, res, publicPath, preloadFiles)
        : defaultPushAssets(preloadFiles, shouldPush, publicPath, options);

      // Pass with single Link header
      // https://blog.cloudflare.com/http-2-server-push-with-multiple-assets-per-link-header
      // https://www.w3.org/Protocols/9707-link-header.html
      if (links.length > 0) {
        res.setHeader('Link', links.join(', '));
      }
    }

    if (options.render.csp && cspScriptSrcHashes) {
      const { allowedSources, policies } = options.render.csp;
      const cspHeader = options.render.csp.reportOnly ? 'Content-Security-Policy-Report-Only' : 'Content-Security-Policy';

      res.setHeader(cspHeader, getCspString({ cspScriptSrcHashes, allowedSources, policies, isDev: options.dev }));
    }

    // Send response
    res.setHeader('Content-Type', 'text/html; charset=utf-8');
    res.setHeader('Accept-Ranges', 'none'); // #3870
    res.setHeader('Content-Length', Buffer.byteLength(html));
    res.end(html, 'utf8');
    await nuxt.callHook('render:routeDone', url, result, context);
    return html
  } catch (err) {
    if (context && context.redirected) {
      consola.error(err);
      return err
    }

    if (err.name === 'URIError') {
      err.statusCode = 400;
    }
    next(err);
  }
};

const defaultPushAssets = (preloadFiles, shouldPush, publicPath, options) => {
  if (shouldPush && options.dev) {
    consola.warn('http2.shouldPush is deprecated. Use http2.pushAssets function');
  }

  const links = [];
  preloadFiles.forEach(({ file, asType, fileWithoutQuery, modern }) => {
    // By default, we only preload scripts or css
    if (!shouldPush && asType !== 'script' && asType !== 'style') {
      return
    }

    // User wants to explicitly control what to preload
    if (shouldPush && !shouldPush(fileWithoutQuery, asType)) {
      return
    }

    const { crossorigin } = options.build;
    const cors = `${crossorigin ? ` crossorigin=${crossorigin};` : ''}`;
    const ref = modern ? 'modulepreload' : 'preload';

    links.push(`<${publicPath}${file}>; rel=${ref};${cors} as=${asType}`);
  });
  return links
};

const getCspString = ({ cspScriptSrcHashes, allowedSources, policies, isDev }) => {
  const joinedHashes = cspScriptSrcHashes.join(' ');
  const baseCspStr = `script-src 'self'${isDev ? ` 'unsafe-eval'` : ''} ${joinedHashes}`;

  if (Array.isArray(allowedSources)) {
    return `${baseCspStr} ${allowedSources.join(' ')}`
  }

  const policyObjectAvailable = typeof policies === 'object' && policies !== null && !Array.isArray(policies);

  if (policyObjectAvailable) {
    const transformedPolicyObject = transformPolicyObject(policies, cspScriptSrcHashes);

    return Object.entries(transformedPolicyObject).map(([k, v]) => `${k} ${v.join(' ')}`).join('; ')
  }

  return baseCspStr
};

const transformPolicyObject = (policies, cspScriptSrcHashes) => {
  const userHasDefinedScriptSrc = policies['script-src'] && Array.isArray(policies['script-src']);

  const additionalPolicies = userHasDefinedScriptSrc ? policies['script-src'] : [];

  // Self is always needed for inline-scripts, so add it, no matter if the user specified script-src himself.
  const hashAndPolicyList = cspScriptSrcHashes.concat(`'self'`, additionalPolicies);

  return { ...policies, 'script-src': hashAndPolicyList }
};

const errorMiddleware = ({ resources, options }) => async function errorMiddleware(err, req, res, next) {
  // ensure statusCode, message and name fields

  const error = {
    statusCode: err.statusCode || 500,
    message: err.message || 'Nuxt Server Error',
    name: !err.name || err.name === 'Error' ? 'NuxtServerError' : err.name
  };

  const sendResponse = (content, type = 'text/html') => {
    // Set Headers
    res.statusCode = error.statusCode;
    res.statusMessage = error.name;
    res.setHeader('Content-Type', type + '; charset=utf-8');
    res.setHeader('Content-Length', Buffer.byteLength(content));
    res.setHeader('Cache-Control', 'no-cache, no-store, max-age=0, must-revalidate');

    // Send Response
    res.end(content, 'utf-8');
  };

  // Check if request accepts JSON
  const hasReqHeader = (header, includes) =>
    req.headers[header] && req.headers[header].toLowerCase().includes(includes);
  const isJson =
    hasReqHeader('accept', 'application/json') ||
    hasReqHeader('user-agent', 'curl/');

  // Use basic errors when debug mode is disabled
  if (!options.debug) {
    // We hide actual errors from end users, so show them on server logs
    if (err.statusCode !== 404) {
      consola.error(err);
    }
    // Json format is compatible with Youch json responses
    const json = {
      status: error.statusCode,
      message: error.message,
      name: error.name
    };
    if (isJson) {
      sendResponse(JSON.stringify(json, undefined, 2), 'text/json');
      return
    }
    const html = resources.errorTemplate(json);
    sendResponse(html);
    return
  }

  const errorFull = err instanceof Error
    ? err
    : typeof err === 'string'
      ? new Error(err)
      : new Error(err.message || JSON.stringify(err));

  errorFull.name = error.name;
  errorFull.statusCode = error.statusCode;
  errorFull.stack = err.stack || undefined;

  // Show stack trace
  const youch = new Youch(
    errorFull,
    req,
    readSourceFactory({
      srcDir: options.srcDir,
      rootDir: options.rootDir,
      buildDir: options.buildDir
    }),
    options.router.base,
    true
  );
  if (isJson) {
    const json = await youch.toJSON();
    sendResponse(JSON.stringify(json, undefined, 2), 'text/json');
    return
  }

  const html = await youch.toHTML();
  sendResponse(html);
};

const readSourceFactory = ({ srcDir, rootDir, buildDir }) => async function readSource(frame) {
  // Remove webpack:/// & query string from the end
  const sanitizeName = name => name ? name.replace('webpack:///', '').split('?')[0] : null;
  frame.fileName = sanitizeName(frame.fileName);

  // Return if fileName is unknown
  if (!frame.fileName) {
    return
  }

  // Possible paths for file
  const searchPath = [
    srcDir,
    rootDir,
    path.join(buildDir, 'dist', 'server'),
    buildDir,
    process.cwd()
  ];

  // Scan filesystem for real source
  for (const pathDir of searchPath) {
    const fullPath = path.resolve(pathDir, frame.fileName);
    const source = await fs.readFile(fullPath, 'utf-8').catch(() => null);
    if (source) {
      frame.contents = source;
      frame.fullPath = fullPath;
      if (path.isAbsolute(frame.fileName)) {
        frame.fileName = path.relative(rootDir, fullPath);
      }
      return
    }
  }
};

let RANDOM_PORT = '0';

class Listener {
  constructor({ port, host, socket, https, app, dev, baseURL }) {
    // Options
    this.port = port;
    this.host = host;
    this.socket = socket;
    this.https = https;
    this.app = app;
    this.dev = dev;
    this.baseURL = baseURL;

    // After listen
    this.listening = false;
    this._server = null;
    this.server = null;
    this.address = null;
    this.url = null;
  }

  async close() {
    // Destroy server by forcing every connection to be closed
    if (this.server && this.server.listening) {
      await this.server.destroy();
      consola.debug('server closed');
    }

    // Delete references
    this.listening = false;
    this._server = null;
    this.server = null;
    this.address = null;
    this.url = null;
  }

  computeURL() {
    const address = this.server.address();
    if (!this.socket) {
      switch (address.address) {
        case '127.0.0.1': this.host = 'localhost'; break
        case '0.0.0.0': this.host = ip.address(); break
      }
      this.port = address.port;
      this.url = `http${this.https ? 's' : ''}://${this.host}:${this.port}${this.baseURL}`;
      return
    }
    this.url = `unix+http://${address}`;
  }

  async listen() {
    // Prevent multi calls
    if (this.listening) {
      return
    }

    // Initialize underlying http(s) server
    const protocol = this.https ? https : http;
    const protocolOpts = typeof this.https === 'object' ? [this.https] : [];
    this._server = protocol.createServer.apply(protocol, protocolOpts.concat(this.app));

    // Call server.listen
    // Prepare listenArgs
    const listenArgs = this.socket ? { path: this.socket } : { host: this.host, port: this.port };
    listenArgs.exclusive = false;

    // Call server.listen
    try {
      this.server = await new Promise((resolve, reject) => {
        this._server.on('error', error => reject(error));
        const s = this._server.listen(listenArgs, error => error ? reject(error) : resolve(s));
      });
    } catch (error) {
      return this.serverErrorHandler(error)
    }

    // Enable destroy support
    enableDestroy(this.server);
    pify(this.server.destroy);

    // Compute listen URL
    this.computeURL();

    // Set this.listening to true
    this.listening = true;
  }

  async serverErrorHandler(error) {
    // Detect if port is not available
    const addressInUse = error.code === 'EADDRINUSE';

    // Use better error message
    if (addressInUse) {
      const address = this.socket || `${this.host}:${this.port}`;
      error.message = `Address \`${address}\` is already in use.`;

      // Listen to a random port on dev as a fallback
      if (this.dev && !this.socket && this.port !== RANDOM_PORT) {
        consola.warn(error.message);
        consola.info('Trying a random port...');
        this.port = RANDOM_PORT;
        await this.close();
        await this.listen();
        RANDOM_PORT = this.port;
        return
      }
    }

    // Throw error
    throw error
  }
}

const createTimingMiddleware = options => (req, res, next) => {
  if (res.timing) {
    consola.warn('server-timing is already registered.');
  }
  res.timing = new ServerTiming();

  if (options && options.total) {
    res.timing.start('total', 'Nuxt Server Time');
  }

  onHeaders(res, () => {
    res.timing.end('total');

    if (res.timing.headers.length > 0) {
      res.setHeader(
        'Server-Timing',
        []
          .concat(res.getHeader('Server-Timing') || [])
          .concat(res.timing.headers)
          .join(', ')
      );
    }
    res.timing.clear();
  });

  next();
};

class ServerTiming extends utils.Timer {
  constructor(...args) {
    super(...args);
    this.headers = [];
  }

  end(...args) {
    const time = super.end(...args);
    if (time) {
      this.headers.push(this.formatHeader(time));
    }
    return time
  }

  clear() {
    super.clear();
    this.headers.length = 0;
  }

  formatHeader(time) {
    const desc = time.description ? `;desc="${time.description}"` : '';
    return `${time.name};dur=${time.duration}${desc}`
  }
}

class Server {
  constructor(nuxt) {
    this.nuxt = nuxt;
    this.options = nuxt.options;

    this.globals = utils.determineGlobals(nuxt.options.globalName, nuxt.options.globals);

    this.publicPath = utils.isUrl(this.options.build.publicPath)
      ? this.options.build._publicPath
      : this.options.build.publicPath;

    // Runtime shared resources
    this.resources = {};

    // Will be set after listen
    this.listeners = [];

    // Create new connect instance
    this.app = connect();

    // Close hook
    this.nuxt.hook('close', () => this.close());

    // devMiddleware placeholder
    if (this.options.dev) {
      this.nuxt.hook('server:devMiddleware', (devMiddleware) => {
        this.devMiddleware = devMiddleware;
      });
    }
  }

  async ready() {
    if (this._readyCalled) {
      return this
    }
    this._readyCalled = true;

    await this.nuxt.callHook('render:before', this, this.options.render);

    // Initialize vue-renderer
    const { VueRenderer } = await Promise.resolve(require('@nuxt/vue-renderer'));

    this.serverContext = new ServerContext(this);
    this.renderer = new VueRenderer(this.serverContext);
    await this.renderer.ready();

    // Setup nuxt middleware
    await this.setupMiddleware();

    // Call done hook
    await this.nuxt.callHook('render:done', this);

    return this
  }

  async setupMiddleware() {
    // Apply setupMiddleware from modules first
    await this.nuxt.callHook('render:setupMiddleware', this.app);

    // Compression middleware for production
    if (!this.options.dev) {
      const { compressor } = this.options.render;
      if (typeof compressor === 'object') {
        // If only setting for `compression` are provided, require the module and insert
        const compression = this.nuxt.resolver.requireModule('compression');
        this.useMiddleware(compression(compressor));
      } else if (compressor) {
        // Else, require own compression middleware if compressor is actually truthy
        this.useMiddleware(compressor);
      }
    }

    if (this.options.server.timing) {
      this.useMiddleware(createTimingMiddleware(this.options.server.timing));
    }

    // For serving static/ files to /
    const staticMiddleware = serveStatic(
      path.resolve(this.options.srcDir, this.options.dir.static),
      this.options.render.static
    );
    staticMiddleware.prefix = this.options.render.static.prefix;
    this.useMiddleware(staticMiddleware);

    // Serve .nuxt/dist/client files only for production
    // For dev they will be served with devMiddleware
    if (!this.options.dev) {
      const distDir = path.resolve(this.options.buildDir, 'dist', 'client');
      this.useMiddleware({
        path: this.publicPath,
        handler: serveStatic(
          distDir,
          this.options.render.dist
        )
      });
    }

    // Dev middleware
    if (this.options.dev) {
      this.useMiddleware((req, res, next) => {
        if (!this.devMiddleware) {
          return next()
        }
        this.devMiddleware(req, res, next);
      });

      // open in editor for debug mode only
      if (this.options.debug) {
        this.useMiddleware({
          path: '__open-in-editor',
          handler: launchMiddleware(this.options.editor)
        });
      }
    }

    // Add user provided middleware
    for (const m of this.options.serverMiddleware) {
      this.useMiddleware(m);
    }

    // Graceful 404 error handler
    const { fallback } = this.options.render;
    if (fallback) {
      // Dist files
      if (fallback.dist) {
        this.useMiddleware({
          path: this.publicPath,
          handler: servePlaceholder(fallback.dist)
        });
      }

      // Other paths
      if (fallback.static) {
        this.useMiddleware({
          path: '/',
          handler: servePlaceholder(fallback.static)
        });
      }
    }

    // Finally use nuxtMiddleware
    this.useMiddleware(nuxtMiddleware({
      options: this.options,
      nuxt: this.nuxt,
      renderRoute: this.renderRoute.bind(this),
      resources: this.resources
    }));

    // Apply errorMiddleware from modules first
    await this.nuxt.callHook('render:errorMiddleware', this.app);

    // Error middleware for errors that occurred in middleware that declared above
    this.useMiddleware(errorMiddleware({
      resources: this.resources,
      options: this.options
    }));
  }

  useMiddleware(middleware) {
    let handler = middleware.handler || middleware;

    // Resolve handler setup as string (path)
    if (typeof handler === 'string') {
      try {
        const requiredModuleFromHandlerPath = this.nuxt.resolver.requireModule(handler);

        // In case the "handler" is not derived from an object but is a normal string, another object with
        // path and handler could be the result

        // If the required module has handler, treat the module as new "middleware" object
        if (requiredModuleFromHandlerPath.handler) {
          middleware = requiredModuleFromHandlerPath;
        }

        handler = requiredModuleFromHandlerPath.handler || requiredModuleFromHandlerPath;
      } catch (err) {
        consola.error(err);
        // Throw error in production mode
        if (!this.options.dev) {
          throw err
        }
      }
    }

    // Resolve path
    const path = (
      (middleware.prefix !== false ? this.options.router.base : '') +
      (typeof middleware.path === 'string' ? middleware.path : '')
    ).replace(/\/\//g, '/');

    // Use middleware
    this.app.use(path, handler);
  }

  renderRoute() {
    return this.renderer.renderRoute.apply(this.renderer, arguments)
  }

  loadResources() {
    return this.renderer.loadResources.apply(this.renderer, arguments)
  }

  renderAndGetWindow(url, opts = {}) {
    return renderAndGetWindow(url, opts, {
      loadedCallback: this.globals.loadedCallback,
      ssr: this.options.render.ssr,
      globals: this.globals
    })
  }

  async listen(port, host, socket) {
    // Ensure nuxt is ready
    await this.nuxt.ready();

    // Create a new listener
    const listener = new Listener({
      port: isNaN(parseInt(port)) ? this.options.server.port : port,
      host: host || this.options.server.host,
      socket: socket || this.options.server.socket,
      https: this.options.server.https,
      app: this.app,
      dev: this.options.dev,
      baseURL: this.options.router.base
    });

    // Listen
    await listener.listen();

    // Push listener to this.listeners
    this.listeners.push(listener);

    await this.nuxt.callHook('listen', listener.server, listener);

    return listener
  }

  async close() {
    if (this.__closed) {
      return
    }
    this.__closed = true;

    await Promise.all(this.listeners.map(l => l.close()));

    this.listeners = [];

    if (typeof this.renderer.close === 'function') {
      await this.renderer.close();
    }

    this.app.removeAllListeners();
    this.app = null;

    for (const key in this.resources) {
      delete this.resources[key];
    }
  }
}

exports.Listener = Listener;
exports.Server = Server;
