// @nuxt/cli/dist/cli.js
/*!
 * @nuxt/cli v2.8.1 (c) 2016-2019

 * - All the amazing contributors
 * Released under the MIT License.
 * Website: https://nuxtjs.org
*/
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

const __chunk_1 = require('./cli-chunk.js');
require('path');
const consola = _interopDefault(require('consola'));
require('minimist');
const exit = _interopDefault(require('exit'));
require('@nuxt/utils');
require('chalk');
require('std-env');
require('wrap-ansi');
require('boxen');
const __chunk_2 = require('./cli-chunk2.js');
require('@nuxt/config');
require('esm');
require('fs-extra');
const fs$1 = _interopDefault(require('fs'));
const execa = _interopDefault(require('execa'));

let _setup = false;

function setup({ dev }) {
  // Apply default NODE_ENV if not provided
  if (!process.env.NODE_ENV) {
    process.env.NODE_ENV = dev ? 'development' : 'production';
  }

  if (_setup) {
    return
  }
  _setup = true;

  // Global error handler
  /* istanbul ignore next */
  process.on('unhandledRejection', (err) => {
    consola.error(err);
  });

  // Exit process on fatal errors
  /* istanbul ignore next */
  consola.addReporter({
    log(logObj) {
      if (logObj.type === 'fatal') {
        const errorMessage = String(logObj.args[0]);
        process.stderr.write(__chunk_2.fatalBox(errorMessage));
        exit(1);
      }
    }
  });

  // Wrap all console logs with consola for better DX
  consola.wrapConsole();
}

async function run(_argv) {
  // Read from process.argv
  const argv = _argv ? Array.from(_argv) : process.argv.slice(2);

  // Check for internal command
  let cmd = await __chunk_1.getCommand(argv[0]);

  // Matching `nuxt` or `nuxt [dir]` or `nuxt -*` for `nuxt dev` shortcut
  if (!cmd && (!argv[0] || argv[0][0] === '-' || fs$1.existsSync(argv[0]))) {
    argv.unshift('dev');
    cmd = await __chunk_1.getCommand('dev');
  }

  // Setup env
  setup({ dev: argv[0] === 'dev' });

  // Try internal command
  if (cmd) {
    return __chunk_1.NuxtCommand.run(cmd, argv.slice(1))
  }

  // Try external command
  try {
    await execa(`nuxt-${argv[0]}`, argv.slice(1), {
      stdout: process.stdout,
      stderr: process.stderr,
      stdin: process.stdin
    });
  } catch (error) {
    if (error.code === 'ENOENT') {
      throw String(`Command not found: nuxt-${argv[0]}`)
    }
    throw String(`Failed to run command \`nuxt-${argv[0]}\`:\n${error}`)
  }
}

const commands = __chunk_1._commands;
const imports = __chunk_1._imports;

exports.NuxtCommand = __chunk_1.NuxtCommand;
exports.loadNuxtConfig = __chunk_1.loadNuxtConfig;
exports.commands = commands;
exports.imports = imports;
exports.run = run;
exports.setup = setup;
