// eslint-plugin-jest/lib/rules/prefer-to-be-null.js
'use strict';

const _require = require('./util'),
      getDocsUrl = _require.getDocsUrl,
      argument = _require.argument,
      argument2 = _require.argument2,
      expectToBeCase = _require.expectToBeCase,
      expectToEqualCase = _require.expectToEqualCase,
      expectNotToEqualCase = _require.expectNotToEqualCase,
      expectNotToBeCase = _require.expectNotToBeCase,
      method = _require.method,
      method2 = _require.method2;

module.exports = {
  meta: {
    docs: {
      url: getDocsUrl(__filename)
    },
    fixable: 'code'
  },

  create(context) {
    return {
      CallExpression(node) {
        const is = expectToBeCase(node, null) || expectToEqualCase(node, null);
        const isNot = expectNotToEqualCase(node, null) || expectNotToBeCase(node, null);

        if (is || isNot) {
          context.report({
            fix(fixer) {
              if (is) {
                return [fixer.replaceText(method(node), 'toBeNull'), fixer.remove(argument(node))];
              }

              return [fixer.replaceText(method2(node), 'toBeNull'), fixer.remove(argument2(node))];
            },

            message: 'Use toBeNull() instead',
            node: is ? method(node) : method2(node)
          });
        }
      }

    };
  }

};
