// eslint-plugin-jest/lib/rules/valid-describe.js
'use strict';

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance"); }

function _iterableToArrayLimit(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

const _require = require('./util'),
      getDocsUrl = _require.getDocsUrl,
      isDescribe = _require.isDescribe,
      isFunction = _require.isFunction;

const isAsync = node => node.async;

const isString = node => node.type === 'Literal' && typeof node.value === 'string' || node.type === 'TemplateLiteral';

const hasParams = node => node.params.length > 0;

const paramsLocation = params => {
  const _params = _slicedToArray(params, 1),
        first = _params[0];

  const last = params[params.length - 1];
  return {
    start: {
      line: first.loc.start.line,
      column: first.loc.start.column
    },
    end: {
      line: last.loc.end.line,
      column: last.loc.end.column
    }
  };
};

module.exports = {
  meta: {
    docs: {
      url: getDocsUrl(__filename)
    }
  },

  create(context) {
    return {
      CallExpression(node) {
        if (isDescribe(node)) {
          if (node.arguments.length === 0) {
            return context.report({
              message: 'Describe requires name and callback arguments',
              loc: node.loc
            });
          }

          const _node$arguments = _slicedToArray(node.arguments, 1),
                name = _node$arguments[0];

          const _node$arguments2 = _slicedToArray(node.arguments, 2),
                callbackFunction = _node$arguments2[1];

          if (!isString(name)) {
            context.report({
              message: 'First argument must be name',
              loc: paramsLocation(node.arguments)
            });
          }

          if (callbackFunction === undefined) {
            return context.report({
              message: 'Describe requires name and callback arguments',
              loc: paramsLocation(node.arguments)
            });
          }

          if (!isFunction(callbackFunction)) {
            return context.report({
              message: 'Second argument must be function',
              loc: paramsLocation(node.arguments)
            });
          }

          if (isAsync(callbackFunction)) {
            context.report({
              message: 'No async describe callback',
              node: callbackFunction
            });
          }

          if (hasParams(callbackFunction)) {
            context.report({
              message: 'Unexpected argument(s) in describe callback',
              loc: paramsLocation(callbackFunction.params)
            });
          }

          if (callbackFunction.body.type === 'BlockStatement') {
            callbackFunction.body.body.forEach(node => {
              if (node.type === 'ReturnStatement') {
                context.report({
                  message: 'Unexpected return statement in describe callback',
                  node
                });
              }
            });
          }
        }
      }

    };
  }

};
